### ISSUE TITLE
Create mixins @base-font-size @base-line-height for reuse

### ISSUE THREAD

**Issue by eduncan911** [atom/atom#11905]
### Prerequisites

* [x] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#check-if-the-problem-shows-up-in-safe-mode)?
* [x] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?
* [x] Did you check the [debugging guide](http://flight-manual.atom.io/hacking-atom/sections/debugging/)?
* [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
* [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
* [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

For more information on how to write a good [bug report](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-bug-report) or [enhancement request](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-enhancement-suggestion), see the `CONTRIBUTING` guide.

### Description

**UPDATE: A simple mixin proposed solution is located at** https://github.com/atom/atom/issues/11905#issuecomment-224336832

**PROBLEM: The functionality for `CMD +` is not accessible by packages and themes.**  Instead, [it is hardcoded to modify `atom-text-editor` element for `font-size` by hardcoding `#px` pixel value](https://github.com/atom/atom/blob/6f1c1fd0e0242b096f4781698066788b26af2fa4/src/workspace-element.coffee#L54).

This does not allow the reuse of this functionality for custom packages and CSS.  The proposed solution noted at https://github.com/atom/atom/issues/11905#issuecomment-224336832 is to move the variable it is modifying to a global mixin, for reuse at this same `atom-text-editor` element and `font-size` attribute.

An example use-case: When resizing the font size with CMD-+ (or CMD-SHIFT-+), only the Editor window is resized - not the TreeView, Tabs, StatusBar, ScrollBar, Dialogs (Find, Packages, etc), or anything else.  

BEFORE: Atom default font...
<img width="1920" alt="screen shot 2016-06-05 at 11 02 13 am" src="https://cloud.githubusercontent.com/assets/104980/15806220/13c5f3b8-2b0d-11e6-95a6-8fba971725ca.png">

AFTER: pressing CMD-+ several times to greatly increase font size...
<img width="1920" alt="screen shot 2016-06-05 at 11 03 34 am" src="https://cloud.githubusercontent.com/assets/104980/15806224/33e268d4-2b0d-11e6-84b1-69bbc841dfab.png">

^- Only the Editor is resized, making the Treeview and Debugger and other dialogs extremely hard to read in screencasts/meetup presentations.

This is really needed, especially for doing Meetups Presentations, Screencasts and even Screen sharing/Code pairing sessions.  Just make a few CMD-+ to blow everything up to enormous size and everyone is happy. 

This was originally mentioned back in November 2015.

> Does anyone know how could I make @font-size change along with zoom levels? E.g. ctrl+/-/scroll-up/down. Thanks!
https://github.com/atom/atom/issues/2530#issuecomment-153580043

If we compare to VSCode, VSCode resizes everything with CMD-+ (Editor window, TreeView, Tabs, StatusBar, Scrollbar, dialogs, and everything else).  For example, here's VSCode before and after:

BEFORE: VSCode default font size...
<img width="1920" alt="screen shot 2016-06-05 at 10 44 53 am" src="https://cloud.githubusercontent.com/assets/104980/15806121/9959af4a-2b0a-11e6-97ef-02f8f6fa23ac.png">

AFTER: pressing CMD-+ a few times blows everything up large.
<img width="1920" alt="screen shot 2016-06-05 at 10 40 08 am" src="https://cloud.githubusercontent.com/assets/104980/15806097/f23fe8be-2b09-11e6-9dac-bdd30a9cca78.png">

^- that's a fullscreen shot at 3840x2400!

In summary, the root issue is that `CMD +` is modifying the base stylesheet directly.  We cannot access this functionality with custom packages and custom CSS.  If a mixin was used instead, simple CSS overrides for the user could reuse them, as noted above.

### Steps to Reproduce

1. Open a code file
2. Press CMD-+ a few times

**Expected behavior:** When pressing CMD-+, I expect a global mixin or some other variable to access for custom packages and custom CSS.

**Actual behavior:** There is no global mixin - it is hardcoded to change the base stylesheet.

### Versions

    $ atom --version
    Atom    : 1.8.0
    Electron: 0.36.8
    Chrome  : 47.0.2526.110
    Node    : 5.1.1

**Comment by eduncan911** [comment]
Talking with @50Wliu on slack, I found this:

https://github.com/atom/atom/blob/2cfc4ee4e6f5f6b9bbc795411839a81c9679db5c/src/workspace.coffee#L738
```
  # Increase the editor font size by 1px.
  increaseFontSize: ->
    @config.set("editor.fontSize", @config.get("editor.fontSize") + 1)

  # Decrease the editor font size by 1px.
  decreaseFontSize: ->
    fontSize = @config.get("editor.fontSize")
    @config.set("editor.fontSize", fontSize - 1) if fontSize > 1

  # Restore to the window's original editor font size.
  resetFontSize: ->
    if @originalFontSize
      @config.set("editor.fontSize", @originalFontSize)

  subscribeToFontSize: ->
    @config.onDidChange 'editor.fontSize', ({oldValue}) =>
      @originalFontSize ?= oldValue
```

I also see `CMD-SHIFT-+` isn't a registered key.  

How about a proposal a new function for `CMD-SHIFT-+` and `CMD-SHIFT- -` to change the everything?

For example, thanks to a tip in this https://github.com/atom/atom/issues/2530#issuecomment-153580043 , I was able to create a base style:

```
 @font-size: 1.2rem;

 html,
 body,
 .tree-view,
 .tooltip,
 .tab-bar .tab,
 .find-and-replace .find-meta-container,
 atom-text-editor,
 .btn,
 .status-bar,
 linter-message {
   font-size: @font-size;
 }
```

(leaving out `atom-text-editor` as I'd continue to want CMD-+ to operate on the editor for now, to have some type of dynamic resizing.

**Comment by lee-dohm** [comment]
<kbd>Cmd-+</kbd> and <kbd>Cmd-Shift-+</kbd> are the same key binding since you can't make <kbd>+</kbd> without using <kbd>Shift</kbd>.

<kbd>Cmd-Shift-+</kbd> is already mapped to the `window:increase-font-size` command:

https://github.com/atom/atom/blob/master/keymaps/darwin.cson#L81

The reason why they are equivalent is because key bindings are case-sensitive, so the short way to write <kbd>cmd-shift-b</kbd> was <kbd>cmd-B</kbd>. People found this unintuitive and confusing, so we changed the convention to write things out. You'll still find vestiges of the old convention though in the Atom code base, like this one.

You can use some code in your `init.coffee` to permanently scale the entire UI like this:

```coffee
require('web-frame').setZoomFactor(1.15)
```

Just change the zoom factor number to your desired scale. One could also easily create a package with whatever keybindings you like to change the zoom factor to scale the entire UI. (This also has the benefit that it scales more than just fonts ... which your suggested approach would only change the font size and things that were explicitly based upon the font size. Various UI icons might not be changed at all, leading to a strange, lopsided experience.)

**Comment by eduncan911** [comment]
https://github.com/atom/atom/blob/6f1c1fd0e0242b096f4781698066788b26af2fa4/src/workspace-element.coffee#L54

Per this, this is hardcoding the `px` size of the font.  

How about we use a `less` variable instead?  That way, we inherit that same variable in other customizations - such as the CSS I posted above.

**Comment by lee-dohm** [comment]
> Per this, this is hardcoding the px size of the font.

Have you tried the `setZoomFactor` thing? It really works:

<img width="1971" alt="screen shot 2016-06-07 at 8 42 42 am" src="https://cloud.githubusercontent.com/assets/1038121/15864414/dd34e73e-2c8b-11e6-8e13-bf1356e22dc7.png">

**Comment by lee-dohm** [comment]
It may require you to reload the window. I'm not sure ... I've only used it in my `init.coffee`.

**Comment by eduncan911** [comment]
Yes, `setZoomFactor` somewhat works... But, it is nowhere near optimum.  It does not resize the scrollbar nor the Tab space and a few other things related to dialogs.  That was all I found in the 30 seconds that i tried it.  The CSS I posted above actually captures a bit more and resizes, including the Tabs spacing (though, I'll add scrollbars now that I think about it).

Also, `setZoomFactor` requires a full restart to take affect.  The CSS method, which is the same method used by `CMD +` to resize the editor font, takes effect immediately and does not require a restart.

Additional info about what to change: 

@50Wliu found this, which shows we are directly editing the `stylesheet` to modify the `font-size`.

https://github.com/atom/atom/blob/6f1c1fd0e0242b096f4781698066788b26af2fa4/src/workspace-element.coffee#L54

I think it will come down to two things to resolve this:

- Create global Mixins called `@base-font-size` and maybe `@base-line-height` (coffee?  less?  I am not sure what Atom uses - I only know less but I'm sure this is simple in coffee).
- Modify the existing `/src/workspace-element.coffee`, around line 54 as lined to above, to use the `@base-font-size` and `@base-line-height` mixins above.

That is all that is needed.  Then, us users who want this functionally can simply use some CSS like I posted earlier, and to use the `@base-font-size` and `@base-line-height` global mixins.

Bonus points if a new keyboard shortcut is created to enable it for everyone.  But for now, getting the base changed to use a mixin would be ideal and would allow us to modify the CSS.

If someone could give me an overview/point me to a tutorial about modifying CSS for Atom, I might take a stab at doing this.

**Comment by lee-dohm** [comment]
> requires you to restart Atom to take affect.

I just tried it in the Developer console and it took effect immediately, didn't even have to reload the window. But yes, it doesn't affect anything that isn't part of the DOM, this includes system dialogs, the OS-provided title bar and the Developer Tools themselves. (Of course, you'll have this same problem with a CSS/LESS solution.) I think scroll bars typically aren't overridden by CSS because people prefer an OS-style experience with them. If they were overridden by CSS in a theme, they'll probably be scaled by `setZoomFactor` as well.

I'm not sure what you mean about Tab spacing though.

**Comment by eduncan911** [comment]
>I just tried it in the Developer console and it took effect immediately,

Needing developer console to adjust the zoom wouldn't be a clean solution for the masses.  Most people would want to modify the `init.coffee` script, which requires a full restart of Atom.  

Hence, my proposed solution above at https://github.com/atom/atom/issues/11905#issuecomment-224336832 which is a simple solution that should allow the masses to just tweak our CSS.

>I'm not sure what you mean about Tab spacing though.

Default Atom tab sizing (setZoomFactor and CSS disabled, restarted Atom):
![screen shot 2016-06-07 at 1 27 59 pm](https://cloud.githubusercontent.com/assets/104980/15868103/3d42aa26-2cb4-11e6-8c52-2a499c129191.png)

My CSS solution (setZoomFactor disabled, only my CSS code - takes affect immediately):
![screen shot 2016-06-07 at 1 26 52 pm](https://cloud.githubusercontent.com/assets/104980/15868115/4b1a6242-2cb4-11e6-8958-e93cd345059d.png)

`setZoomFactor` set to 1.5 (my CSS disabled, setZoomFactor used and needing restart):
![screen shot 2016-06-07 at 1 29 22 pm](https://cloud.githubusercontent.com/assets/104980/15868128/5829be42-2cb4-11e6-9e7d-e443c70f70b4.png)

^- That is what all tabs look like after restarting with `setZoomFactor`.  Don't know how you could miss that with `setZoomFactor`.

**Comment by eduncan911** [comment]
What about my proposed solution at https://github.com/atom/atom/issues/11905#issuecomment-224336832 ?

**Comment by lee-dohm** [comment]
> Needing developer console to adjust the zoom wouldn't be a clean solution for the masses. Most people would want to modify the init.coffee script, which requires a full restart of Atom.

I was proving the theory that a package could be written (and thereby a keybinding) where the change would take effect immediately, that one doesn't need to restart Atom. I wasn't suggesting that it was the final solution.

As far as the tab spacing goes, you're probably pushing across the border where Atom switches to "mini" mode for smaller screens. That could be solved with some minor tweaks to the default themes.

> What about my proposed solution at #11905 (comment) ?

Your proposed solution would:

1. Have the exact same drawbacks as the `setZoomFactor` solution
1. Would require possibly large-scale rewriting of just the default themes for it to even start working
1. Would require large-scale rewriting of all Atom themes and potentially even non-theme packages would need to be rewritten for it to work reliably

`setZoomFactor` is:

1. Equivalent to your CSS/LESS solution
1. Doesn't require any changes to themes to start working
1. Requires only minor tweaks for some edge cases in themes (like the tab spacing issue)

If you want a short-term fix for this, figuring out how to make `setZoomFactor` work for you is going to be your best bet.

**Comment by eduncan911** [comment]
Not sure I agree entirely.

Your proposed solution would:

> Have the exact same drawbacks as the setZoomFactor solution
> 1- Have the exact same drawbacks as the setZoomFactor solution

What are the drawbacks for using CSS exactly?  the CSS solution seems to be working, whereas the `setZoomFactor` needs a restart (unless a "package" can hack around that, as you propose - more hacks?).

`setZoomFactor` may work but as you pointed out, it may be "switching to 'mini' mode" as a complete guess - pointing out there's even more work to do  This is starting to smell of a rabbit hole to investigate how to forcefully hack a restart or alike, figure out the "mini mode" if that is even the problem here, and who knows what else would come up.

> 2 - Would require possibly large-scale rewriting of just the default themes for it to even start working

I don't think this is the case.  Maybe I am mis-understanding, but this is simple CSS right? 

Currently it is being hardcoded by directly inserting it into base stylesheet as you can see here: https://github.com/atom/atom/blob/6f1c1fd0e0242b096f4781698066788b26af2fa4/src/workspace-element.coffee#L54 .  

How is that any different than adjusting the value of a mixin variable?  And, that mixin variable is just used for that `atom-text-editor` element located in the same base stylesheet and using it for this `font-size` value?  

It's all re-parsed in the end the same way upon refresh using the existing event hooks that are already in place.  No class names or elements are changed.

> 3- Would require large-scale rewriting of all Atom themes and potentially even non-theme packages would need to be rewritten for it to work reliably

I don't think this is the case either.  It is already being hardcoded/directly inserted into the base stylesheet - my proposal doesn't change any of that (no class name changes, no css attribute changes, nada - 100% backwards compatibility).  

As a matter of fact, I just move the modifications to a global variable for re-usability, starting with only 1 use: the existing `font-size` already being hardcoded/inserted into the DOM.  

Any custom themes and packages out there already would forcefully have to overwrite this same `atom-text-editor` element and `font-size` attribute.  That does not change.

**Comment by simurai** [comment]
> Currently it is being hardcoded by directly inserting it into base stylesheet as you can see here
> How is that any different than adjusting the value of a mixin variable?

There is the [@font-size](https://github.com/atom/atom/blob/master/static/variables/ui-variables.less#L69) variable that themes define and packages use. But it's read-only. Changing that variable at runtime would require all the Less files that use it to be recompiled and injected again. Probably all CSS since source order matters. There were some attempts to do that, but because of [performance concerns](https://github.com/atom/atom/pull/4153#issuecomment-73980380) didn't end up in production. Now CSS custom properties (that just got added to Atom, I think currently in Atom Beta) could be updated at runtime without recompiling. But because it's a different syntax (`var(--font-size)` instead of `@font-size`) it would only work for themes that get updated to use the custom property.

@eduncan911 what theme are you using? Based on the screenshots, looks like One dark? With that theme, you can scale the UI by changing the "Font Size" theme setting. It doesn't scale everything like `setZoomFactor`, but most parts.

![screen shot 2016-06-08 at 10 57 09 pm](https://cloud.githubusercontent.com/assets/378023/15896711/665553fa-2dcc-11e6-9993-e274935b2c63.png)

Personally, I like that `cmd-+/-` only changes the editor, but keeps the UI as is. And only if needed, e.g. meetup, the UI can still be changed in the settings.

Or you can also change it in your `styles.less` file with:

```css
.theme-one-dark-ui {
  font-size: 14px;
}
```

**Comment by simurai** [comment]
_Maybe_ it's possible to create an init script that listens to the `atom-text-editor` font-size changes and then inlines the same value to the `html` element, like `<html style="font-size: 14px;">`, then editor and UI would be synced. But that only works for themes that use the `em` units (like One dark).

**Comment by eduncan911** [comment]
@simurai

> There is the @font-size variable that themes define and packages use. But it's read-only. Changing that variable at runtime would require all the Less files that use it to be recompiled and injected again. Probably all CSS since source order matters. There were some attempts to do that, but because of performance concerns didn't end up in production. Now CSS custom properties (that just got added to Atom, I think currently in Atom Beta) could be updated at runtime without recompiling. But because it's a different syntax (var(--font-size) instead of @font-size) it would only work for themes that get updated to use the custom property.

That sounds... a lot more complicated than I thought!  I would have thought that when the base stylesheet changes (either by user entries + SAVE, or by the `CMD +` action - which currently works) that it parses all CSS, starting with the base, and up the chain.  It seems this is how it works right now.

The goal of this Issue was to create some global `@base-font-size` that is _dynamically changed with some action_.   `CMD +` or some other new shortcut.  It would only be changed for the `atom-text-editor` to use for now, as it would not change any functionality of CSS overrides.  It would just allow someone like me to tweak my own CSS, to use that `@base-font-size`. Later on, theme authors could elect to use it if they wish - an option to have a variable dynamically changed by some shortcut key.  Like `CMD +`.

>eduncan911 what theme are you using? Based on the screenshots, looks like One dark? 

UI Theme: One Dark, yep
Syntax Theme: Tesla Atom (for those that are interested in my "I left Solarized!" theme)

>Personally, I like that cmd-+/- only changes the editor, but keeps the UI as is. And only if needed, e.g. meetup, the UI can still be changed in the settings.

I agree, in that this isn't the feature I am talking about developing - changing the entire UI with `CMD +` (though, I really LOVED that feature of VSCode when I saw it did it).  I am only talking about creating some sort of dynamically-changing variable that is tied to an Action, that we can link to in CSS stylesheets.

>Maybe it's possible to create an init script that listens to the atom-text-editor font-size changes and then inlines the same value to the html element, like <html style="font-size: 14px;">, then editor and UI would be synced. But that only works for themes that use the em units (like One dark).

that's one idea to directly address the example I gave in the OP.  Didn't want to go that far as to build in the actual change of the UI - just wanted to expose a dynamically-changing variable that we could hook into.  Also, there may be other elements that need changing (though, on the `html` element is pretty broad).

-- 

**Alternative proposal:**

What about creating an init script that listens to the `atom-text-editor` font-size changes (and maybe line-height?) and then modifies that `var(--font-size)` you mention?  

If it can be changed at Runtime with an action (e.g. `CMD +`), and it allows us to use it simply by inserting it into the global Stylesheet (in order to bind it to whatever elements we want, or don't want, to dynamically change), I think that's the winner.

**Comment by simurai** [comment]
> The goal of this Issue was to create some global `@base-font-size` that is _dynamically changed with some action_.

That would be nice, but as mentioned above, due to performance concerns, it isn't something that Atom currently supports. Some packages tried to "hack" it.. like the `base16-syntax` theme switcher. I think this part does the [updating of the variables](https://github.com/Alchiadus/base16-syntax/blob/master/lib/base16.coffee#L26-L40).

> Alternative proposal:

Yeah, that's something we could look into. Although it needs to be done on a per theme/package basis. There are probably lots of "abandoned" themes that even if we would send a PR, the theme author wouldn't bother about merging it. So for an official support of all themes, it has to wait for Atom `2.0`, which there aren't any plans yet when that might be.

**Comment by simurai** [comment]
Here a custom init script and keybinding to increase/decrease the UI font-size:

```coffee
# init.coffee
UIFontSize = 12
atom.commands.add 'atom-workspace',
  'ui:increase-font-size': ->
    UIFontSize += 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
  'ui:decrease-font-size': ->
    UIFontSize -= 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
```

```cson
# keymap.cson
'atom-workspace':
  'ctrl-+': 'ui:increase-font-size'
  'ctrl--': 'ui:decrease-font-size'
```

So `cmd-+/-` is to change the editor font size and `ctrl-+/-` the UI. Doesn't work with all UI themes, below an example with One dark:

![font-size](https://cloud.githubusercontent.com/assets/378023/15927753/5e29d1bc-2e7e-11e6-9d0c-8acb47de6442.gif)

**Comment by eduncan911** [comment]
@simurai Most Excellent!!! 👍 

Though, I had to use an underscore for the decrease-font-size (maybe because of my PC keyboard on OSX):

```
# keymap.cson
'atom-workspace':
  'ctrl-+': 'ui:increase-font-size'
  'ctrl-_': 'ui:decrease-font-size'
```

Further tweaking...  This is my first attempt at touching a line of `coffee script`:

Is it not possible to call the `updateGlobalTextEditorStyleSheet()` function to update the Editor's font size?

```
# init.coffee
UIFontSize = 12
atom.commands.add 'atom-workspace',
  'ui:increase-font-size': ->
    UIFontSize += 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
    @updateGlobalTextEditorStyleSheet()
  'ui:decrease-font-size': ->
    UIFontSize -= 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
    @updateGlobalTextEditorStyleSheet()
```

Code is located here: https://github.com/atom/atom/blob/6f1c1fd0e0242b096f4781698066788b26af2fa4/src/workspace-element.coffee#L54

I'm aware that this could get things outta wack if I start from different sizes.  I can reset things.  :)

**Comment by simurai** [comment]
> The idea is I could use CTRL-SHIFT-+ to scale everything at once

If you want it to be the same all the time, you could hook into the config changes and add the same value to the `html` element. Then UI and editor font-size are always synced:

```coffee
# init.coffee

# set initial value
document.documentElement.style.fontSize = atom.config.get('editor.fontSize') + 'px'

# update value
atom.config.onDidChange 'editor.fontSize', ({newValue}) ->
  document.documentElement.style.fontSize = newValue + 'px'
```

**Comment by eduncan911** [comment]
And... we have a winner!  

❤️  x 100!

--

I think we can close this issue, since `var(--font-size)` kind of stuff is in beta and may allow the customization we were seeking here.

**Comment by michaelmcandrew** [comment]
Thanks for the script!

> I'm aware that this could get things outta wack if I start from different sizes. I can reset things. :)

Agreed - here is a reset
```coffeescript
# init.coffee
UIFontSize = 12
atom.commands.add 'atom-workspace',
  'ui:increase-font-size': ->
    UIFontSize += 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
    @updateGlobalTextEditorStyleSheet()
  'ui:decrease-font-size': ->
    UIFontSize -= 1
    document.documentElement.style.fontSize = UIFontSize + 'px'
    @updateGlobalTextEditorStyleSheet()
  'ui:reset-font-size': ->
    UIFontSize = 12
    document.documentElement.style.fontSize = UIFontSize + 'px'
    @updateGlobalTextEditorStyleSheet()
```
```coffeescript
# keymap.cson
'atom-workspace':
  'ctrl-+': 'ui:increase-font-size'
  'ctrl-_': 'ui:decrease-font-size'
  'ctrl-)': 'ui:reset-font-size'

```

**Comment by NullVoxPopuli** [comment]
@eduncan911's code works flawlessly!

**Comment by harshbhatt** [comment]
Posting on an old thread, but is there any way @eduncan911's script can be ported/modified to work for other themes? Not too familiar with how the elements are picked up/how they are defined differently by the themes. Thanks for the script though!
